### PR TITLE
big query record type changes

### PR DESCRIPTION
**Pull request by Sridhar-Durgam2022**
Read part -> record type changes

### CODE REVIEW

File: src/main/java/com/google/cloud/flink/bigquery/ArrowRowDataDeserializationSchema.java [hunk 1]
```diff
@@ -58,8 +64,37 @@ public ArrowRowDataDeserializationSchema(
...
-        ArrowToRowDataConverters.createRowConverter(
-            rowType, readSessionFieldNames, selectedFieldList);
+        ArrowToRowDataConverters.createRowConverter(rowType, readSessionFieldNames);
+  }
+
+  /**
+   * @param rowType
+   * @param selectedFieldList
+   * @param arrowFieldList
+   * @return
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/ArrowRowDataDeserializationSchema.java]: Removed

File: src/main/java/com/google/cloud/flink/bigquery/ArrowRowDataDeserializationSchema.java [hunk 2]
```diff
@@ -58,8 +64,37 @@ public ArrowRowDataDeserializationSchema(
...
     this.runtimeConverter =
-        ArrowToRowDataConverters.createRowConverter(
-            rowType, readSessionFieldNames, selectedFieldList);
+        ArrowToRowDataConverters.createRowConverter(rowType, readSessionFieldNames);
+  }
+
+  /**
+   * @param rowType
+   * @param selectedFieldList
+   * @param arrowFieldList
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/ArrowRowDataDeserializationSchema.java]: Removed

File: src/main/java/com/google/cloud/flink/bigquery/ArrowFormatFactory.java [hunk 3]
```diff
@@ -42,23 +43,20 @@
...
-              formatOptionsString.indexOf("selectedFields="),
-              formatOptionsString.indexOf(" ", formatOptionsString.indexOf("selectedFields=")));
-      if (this.selectedFields.endsWith(",")) {
-        this.selectedFields =
-            selectedFields.substring(0, selectedFields.length() - 1).replace("selectedFields=", "");
-      }
+          selectedFields.substring(0, selectedFields.length() - 1).replace("selectedFields=", "");
     }
     List<String> selectedFieldList = new ArrayList<String>();
+    List<String> arrowFieldList = new ArrayList<String>();
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/ArrowFormatFactory.java]: Fixed

File: src/main/java/com/google/cloud/flink/bigquery/ArrowRowDataDeserializationSchema.java [hunk 4]
```diff
@@ -58,8 +64,37 @@ public ArrowRowDataDeserializationSchema(
     this.arrowSchemaJson = arrowSchema.toJson().toString();
     this.nestedSchema = ArrowDeserializationSchema.forGeneric(arrowSchemaJson, typeInfo);
     this.runtimeConverter =
-        ArrowToRowDataConverters.createRowConverter(
-            rowType, readSessionFieldNames, selectedFieldList);
+        ArrowToRowDataConverters.createRowConverter(rowType, readSessionFieldNames);
+  }
+
+  /**
+   * @param rowType
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/ArrowRowDataDeserializationSchema.java]: Removed

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/cloud/flink/bigquery/ArrowFormatFactory.java [hunk 5]
```diff
@@ -42,23 +43,19 @@
...
-              formatOptionsString.indexOf("selectedFields="),
-              formatOptionsString.indexOf(" ", formatOptionsString.indexOf("selectedFields=")));
-      if (this.selectedFields.endsWith(",")) {
-        this.selectedFields =
-            selectedFields.substring(0, selectedFields.length() - 1).replace("selectedFields=", "");
-      }
+          selectedFields.substring(0, selectedFields.length() - 1).replace("selectedFields=", "");
     }
     List<String> selectedFieldList = new ArrayList<String>();
+    List<String> arrowFieldList = Arrays.asList(options.get("arrowFields").split(","));
```
[the remark this answers is not in the document]
Comment (quantiphi-anitha) [reply src/main/java/com/google/cloud/flink/bigquery/ArrowFormatFactory.java:54]: selectedFields are the comma separated fields names that are passing to connector, that need to be selected from the given read table.
arrowFields are comma separated field names that are getting from arrow Schema of read session.

There are few situations that the order of columns in arrow schema are different when it's get from read session. In this case, the root.load() fails with cast issue like String not able to convert to Long.
So we are getting this arrow fields to rearrange the producedDataType to generate arrow Schema.
Same time, we cannot use the arrow schema getting from read session, if the Flink table column names are different from the read table.

So, using selected fields, rowtype and arrow fields to rearrange the producedDataType.

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/cloud/flink/bigquery/BigQueryDynamicTableFactory.java [hunk 6]
```diff
@@ -207,6 +217,17 @@ private ArrayList<String> getReadStreamNames(ReadableConfig options) {
       for (ReadStream stream : readsessionList) {
         readStreamNames.add(stream.getName());
       }
+
+      Schema arrowSchema =
+          MessageSerializer.deserializeSchema(
+              new ReadChannel(
+                  new ByteArrayReadableSeekableByteChannel(
+                      readSession.getArrowSchema().getSerializedSchema().toByteArray())));
+      arrowSchema.getFields().stream()
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/BigQueryDynamicTableFactory.java:226]: changed

File: src/main/java/com/google/cloud/flink/bigquery/BigQueryDynamicTableFactory.java [hunk 7]
```diff
@@ -154,12 +159,17 @@ public DynamicTableSource createDynamicTableSource(Context context) {
         throw new IllegalArgumentException(exceptionString);
       }
     }
-
+    ArrayList<String> readStreams = getReadStreamNames(options);
+    context
+        .getCatalogTable()
+        .getOptions()
+        .put("arrowFields", arrowFields.substring(0, arrowFields.length() - 1));
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/BigQueryDynamicTableFactory.java:166]: addressed

File: src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java [hunk 8]
```diff
@@ -45,20 +45,20 @@
...
+  private TypeInformation<RowData> typeInfo;
   private VectorSchemaRoot root;
   private VectorLoader loader;
-  List<FieldVector> vectors = new ArrayList<>();
+  private List<FieldVector> vectors = new ArrayList<>();
   private Schema schema;
-  private final Class<T> recordClazz;
-  private Schema readSessionSchema;
+  private Class<T> recordClazz;
+  private String schemaJsonString;
```
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java:54]: Schema is not serializable, so converting to Schema to String and vice versa

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java:54]: addressed

File: src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java [hunk 9]
```diff
@@ -88,7 +86,9 @@ public T deserialize(byte[] responseByteMessage) throws IOException {
     return (T) root;
   }
 
-  private void initializeArrow() {
+  private void initializeArrow() throws IOException {
+    Schema schema = getSchema(schemaJsonString);
+
     Preconditions.checkNotNull(schema);
```
Comment (davidrabinowitz) [inline src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java:92]: getSchema cannot return null, so the check is redundant

File: src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java [hunk 10]
```diff
@@ -113,6 +113,17 @@ public TypeInformation<T> getProducedType() {
     return (TypeInformation<T>) typeInfo;
   }
 
+  private Schema getSchema(String schemaJson) {
+    Schema schema = null;
+    try {
+      schema = Schema.fromJSON(schemaJson);
+    } catch (IOException e) {
+      logger.error("Error while converting to Schema from jsonString");
+      throw new FlinkBigQueryException("Error while converting to Schema from jsonString");
```
Comment (davidrabinowitz) [inline src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java:122]: please add the thrown exception as the cause of the new one: ` throw new FlinkBigQueryException("Error while converting to Schema from jsonString", e);`

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java (hunk 9 above)
Comment (Sridhar-Durgam2022) [reply src/main/java/com/google/cloud/flink/bigquery/ArrowDeserializationSchema.java]: removed the redundant check

**Review by davidrabinowitz — APPROVED** [review]

**Pull request merged**